### PR TITLE
Fix 435

### PR DESCRIPTION
**Pull request by redboltz**

### CODE REVIEW

**Comment by redboltz** [comment]
@jonesmz , this PR fixes `std::shared_ptr<endpoint_t>` resource management as the original design. When I introduced zero copy mechanism on v5.0.0, this bug has been en-bugged.

NOTE: Original design 

On the server side such as broker, pass the lambda expression that captures `shared_ptr<endpoint_t>` to `start_session()` as `async_handler_t` as follows:

```cpp
        ep.start_session(
            [sp] // keeping ep's lifetime as sp until session finished
            (boost::system::error_code const& /*ec*/) {
            }
        );
```

If ep is closed, then pointee of `shared_ptr<endpoint_t>` (ep) is deleted, so we don't need to clear handlers.

I will merge the PR. After I merged, please rebase #430 from the merged master.
Then, let's consider the lifetime issue again.

File: test/connect.cpp [hunk 1]
```diff
@@ -566,13 +566,9 @@ BOOST_AUTO_TEST_CASE( noclean ) {
...
-                        // If clean session is not provided, than there will be a session present
-                        // if there was ever a previous connection, even if clean session was provided
-                        // on the previous connection.
-                        // This is because MQTTv5 change the semantics of the flag to "clean start"
-                        // such that it only effects the start of the session.
-                        // Post Session cleanup is handled with a timer, not with the  clean session flag.
-                        BOOST_TEST(sp == true);
+                        // The previous connection is not set Session Expiry Interval.
+                        // That means session state is cleared on close.
+                        BOOST_TEST(sp == false);
```
Comment (jonesmz) [inline test/connect.cpp]: I thought that if session expiry interval was not set, that means indefinite lifetime of the session data?

File: test/connect.cpp [hunk 2]
```diff
@@ -603,7 +599,22 @@ BOOST_AUTO_TEST_CASE( noclean ) {
...
-                    c->connect();
+                    switch (c->get_protocol_version()) {
+                    case MQTT_NS::protocol_version::v3_1_1:
+                        c->connect();
+                        break;
+                    case MQTT_NS::protocol_version::v5:
+                        // set session_expiry_interval as infinity.
+                        c->connect(
+                            std::vector<MQTT_NS::v5::property_variant>{
+                                MQTT_NS::v5::property::session_expiry_interval(0xFFFFFFFFUL)
```
Comment (jonesmz) [inline test/connect.cpp]: Is this infinity, or just maximum? It's a uint16_t, right?

Comment (jonesmz) [reply test/connect.cpp]: Ah, it's actually infinity.

3.1.2.11.2 Session Expiry Interval
17 (0x11) Byte, Identifier of the Session Expiry Interval.

Followed by the Four Byte Integer representing the Session Expiry Interval in seconds. It is a Protocol Error to include the Session Expiry Interval more than once.

 

If the Session Expiry Interval is absent the value 0 is used. If it is set to 0, or is absent, the Session ends when the Network Connection is closed.

 

If the Session Expiry Interval is 0xFFFFFFFF (UINT_MAX), the Session does not expire.

 

The Client and Server MUST store the Session State after the Network Connection is closed if the Session Expiry Interval is greater than 0 [MQTT-3.1.2-23].

File: test/connect.cpp (hunk 1 above)
Comment (jonesmz) [reply test/connect.cpp]: Nervermind. I looked it up.

> If the Session Expiry Interval is absent the value 0 is used. If it is set to 0, or is absent, the Session ends when the Network Connection is closed.

File: test/resend.cpp [hunk 3]
```diff
@@ -214,10 +231,23 @@ BOOST_AUTO_TEST_CASE( publish_qos1 ) {
...
         c->set_error_handler(
-            [&chk, &c]
+            [&chk, &c, &tim]
             (boost::system::error_code const&) {
                 MQTT_CHK("h_error");
-                c->connect();
+                // TCP level disconnection detecting timing is unpredictable.
+                // Sometimes broker first, sometimes the client (this test) first.
+                // This test assume that the broker detects first, so I set timer.
+                // If client ditect the disconnection first, then reconnect with
```
Comment (jonesmz) [inline test/resend.cpp]: ditect = detect

File: test/resend.cpp [hunk 4]
```diff
@@ -374,10 +400,23 @@ BOOST_AUTO_TEST_CASE( publish_qos2 ) {
...
```
Comment (jonesmz) [inline test/resend.cpp]: Same here

File: test/resend_serialize_ptr_size.cpp [hunk 5]
```diff
@@ -12,6 +12,27 @@ BOOST_AUTO_TEST_SUITE(test_resend_serialize_ptr_size)
 
 using namespace MQTT_NS::literals;
 
+template <typename Client>
+inline void connect_no_clean(Client& c) {
```
Comment (jonesmz) [inline test/resend_serialize_ptr_size.cpp]: This looks like the same code in multiple files.

Can it be put in a header file?

File: test/connect.cpp (hunk 1 above)
Comment (redboltz) [reply test/connect.cpp]: Ok. no problem.

File: test/resend.cpp (hunk 3 above)
Comment (redboltz) [reply test/resend.cpp]: I will fix all of them.

File: test/resend_serialize_ptr_size.cpp (hunk 5 above)
Comment (redboltz) [reply test/resend_serialize_ptr_size.cpp]: Yes I will move it to header file.
I've checked existing header files but not found appropriate one. So I will create `test_util.hpp`.

**Pull request merged**